### ISSUE TITLE
https traffic is not proxied (in windows)

### ISSUE THREAD

**Issue by editedredx** [greim/hoxy#57]
After the error fix from #56 I'm still not able to catch https traffic, firefox is responding with a **Secure Connection Failed** error. Http traffic is working as expected. The code is still the same:

```javascript
var port = 46769;
var proxy = hoxy.createServer({
    certAuthority: {
        key: fs.readFileSync('f:/ssl/key.pem'),
        cert: fs.readFileSync('f:/ssl/cert.pem')
    }
}).listen(port);

proxy.intercept({
    phase: 'request'
}, function(req, resp, cycle) {
    console.log(123);
});
console.log('tunnel through', port);
```

No console logging happens when requesting through https and the error in firefox happens as a timeout. (I've tried it in Chrome with no difference in behaviour)

![schermafbeelding_082115_083519_am](https://cloud.githubusercontent.com/assets/642298/9402646/98da41ce-47df-11e5-91f0-fd5d9b9d2e4b.jpg)

**Comment by greim** [comment]
Can you verify that your `cert` has been added to Firefox's list of trusted authorities? (http://greim.github.io/hoxy/#intercept-https)

Also, can you verify OpenSSL is in your path? (https://github.com/andris9/pem/issues/58)

**Comment by editedredx** [comment]
Creating the ssl files using your examples did not work at first so I created them with other commands, I noticed your creating a certificate using rather specific settings, so that went wrong.
```coffeescript
let { clientKey: key, certificate: cert } = yield create({
    country: 'US',
    state: 'Utah',
    locality: 'Provo',
    organization: 'ACME Tech Inc',
    commonName: serverName,
    serviceKey: rootKey,
    serviceCertificate: rootCert,
    serial: Date.now(),
    days: 500,
})
```
I finally found what went wrong with creating the files usiong your example so I created them that way.

(BTW, adding a second argument to the above function with a callback handling some error logging helped a lot, I'd think adding it would be a good addition
```coffeescript
let { clientKey: key, certificate: cert } = yield create({
    country: 'US',
    state: 'Utah',
    locality: 'Provo',
    organization: 'ACME Tech Inc',
    commonName: serverName,
    serviceKey: rootKey,
    serviceCertificate: rootCert,
    serial: Date.now(),
    days: 500,
}, function(err, data) {
    if(err) console.log('err: ', err);
    console.log(data);
})
```
)

So currently I have created the my-private-root-ca.key.pem and my-private-root-ca.crt.pem exactly via the commands you used. This yields no errors from the above mentioned callback.
I have added the following before creating the hoxy server:
```javascript
// not really needed, openssl can be found on my path
process.env.OPENSSL_BIN = 'C:\\Program Files (x86)\\Git\\bin\\openssl.exe'; 
// this was not found by default in my setup
process.env.OPENSSL_CONF = 'C:\\Program Files (x86)\\Git\\ssl\\openssl.cnf'; 
```

The errors there are solved when I run the scenario now it still does not respond to ssl traffic, the data log returns the certificates and keys and stuff but still no logging from the interceptor nor the page getting a response.
The certificate is added to the firefox trusted authorities.
![schermafbeelding_082415_121853_pm](https://cloud.githubusercontent.com/assets/642298/9437916/48ea91be-4a5a-11e5-9bcd-b12667c3af86.jpg)

**Comment by greim** [comment]
Rats. I was hoping there would be an easy fix :( I'll have to try to replicate this on Windows, which takes a bit of effort since I have to do it in VirtualBox.

**Comment by editedredx** [comment]
Funny thing is, if I do not have the callback function with the error and data logging in it, I get another EADDRNOTAVAIL error.
```
Error: connect EADDRNOTAVAIL
    at exports._errnoException (util.js:746:11)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1010:19)
```
When the callback is there, the error is empty and the data consists of an object with `csr`, `clientKey`, `certificate` and `serviceKey` keys.

**Comment by greim** [comment]
```js
let create = adapt.part(pem.createCertificate)
```

Adding a callback to the `create()` call will likely cause incorrect behavior, since `create()` [returns a promise](https://www.npmjs.com/package/ugly-adapter). To properly listen for errors on it, maybe wrap `yield create(...)` in a try/catch and then re-throw the error from the catch.

**Comment by editedredx** [comment]
Alright, when doing it with a try/catch block I get the `EADDRNOTAVAIL` error, when I try it like this:
```javascript
return create({
    country: 'US',
    state: 'Utah',
    locality: 'Provo',
    organization: 'ACME Tech Inc',
    commonName: serverName,
    serviceKey: rootKey,
    serviceCertificate: rootCert,
    serial: Date.now(),
    days: 500
}).then(function(data) {
    console.log('success', arguments);
    return data;
},function() {
    console.log('err', arguments);
});
```
I get the `EADDRNOTAVAIL` error, when I leave out the `return data;` I get no error. So basically the error is happening after the create call if it receives data.

**Comment by greim** [comment]
Since you're able to replicate this directly, in your environment, in proxy.js line 218, can you try changing:

```js
let serverSocket = net.connect(addr.port, addr.address, () => {
```

...to this:

```js
let serverSocket = net.connect(addr.port, 'localhost', () => {
```

...and see if that fixes the error?

**Comment by editedredx** [comment]
Tried that to no avail... however in the process I did find the answer a bit lower.
On line 235 I added a log of `addr` resulting in the following:
```javascript
{ address: '::', family: 'IPv6', port: 46769 }
```
Not sure if you've tested the ssl methods with IPv6 yet? Anyway, besides that, I don't think the address of `::` is valid it's either `::1` or `localhost` and preferably the latter since that is IPv4 compatible.
so changing the address did the trick!
```javascript
let toServer = http.request({
  host: 'localhost',
  port: addr.port,
  method: fromClient.method,
  path: fullUrl,
  headers: fromClient.headers,
}, fromServer => {
  toClient.writeHead(fromServer.statusCode, fromServer.headers)
  fromServer.pipe(toClient)
})
```
I'm getting https pages now as well as intercepts being handled.

**Comment by editedredx** [comment]
Sorry, accidentally closed the issue while it's not fixed in Hoxy yet...

**Comment by greim** [comment]
Was that in addition to my suggested fix above, or instead of?

**Comment by editedredx** [comment]
I tested it with and without your fix, so far it made no difference so I
think it's not needed.

**Comment by greim** [comment]
Just pushed this fix to master. https://github.com/greim/hoxy/commit/1e7d50ae833d1358bf1cf2e24daad9694d9b3a68 Do you want to verify it works, or are you confident it will solve this issue?

**Comment by editedredx** [comment]
Just verified, https pages coming through and are intercepted!

**Comment by majdamarcin** [comment]
I just installed newest hoxy (3.1.2) and for that code:
```javascript
var hoxy = require('hoxy');
var fs = require('fs');

var PROXY_PORT = 8001;

var proxy = new hoxy.createServer({
	certAuthority: {
		key: fs.readFileSync('./my-private-root-ca.key.pem'),
		cert: fs.readFileSync('./my-private-root-ca.crt.pem')
	}
}).listen(PROXY_PORT, function() {
	console.log('proxy listens on: ' + PROXY_PORT);
});

proxy.intercept('request', function(req, resp, cycle) {
	console.log('req');  
    console.log('resp');
});

proxy.log('error warn debug info');
```

i'm getting errors in terminal:
```bash
28 Aug 08:58:55 - [nodemon] restarting due to changes...
28 Aug 08:58:55 - [nodemon] starting `node run-hoxy.js`
28 Aug 08:58:55 - [nodemon] restarting due to changes...
28 Aug 08:58:55 - [nodemon] restarting due to changes...
28 Aug 08:58:55 - [nodemon] restarting due to changes...
28 Aug 08:58:56 - [nodemon] starting `node run-hoxy.js`
proxy listens on: 8001
INFO: generated fake credentials for localhost
INFO: generated fake credentials for ssl.gstatic.com
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: connect EADDRNOTAVAIL
    at exports._errnoException (util.js:746:11)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1010:19)
28 Aug 08:59:13 - [nodemon] app crashed - waiting for file changes before starting...
```

am i missing something?

**Comment by editedredx** [comment]
3.1.2 does not have the fix in it yet, you need to build the project yourself if you want to test it.

**Comment by majdamarcin** [comment]
Thanks, I will check it.

**Comment by majdamarcin** [comment]
Upgrade resolved problem with EADDRNOTAVAIL, but now I'm getting: 
```javascript
28 Aug 14:53:21 - [nodemon] v1.4.1
28 Aug 14:53:21 - [nodemon] to restart at any time, enter `rs`
28 Aug 14:53:21 - [nodemon] watching: *.*
28 Aug 14:53:21 - [nodemon] starting `node run-hoxy.js`
proProxy API listens on: 8002
proProxy listens on: 8001
INFO: generated fake credentials for www.google.pl
INFO: generated fake credentials for debug.mms.localhost
INFO: generated fake credentials for ssl.gstatic.com
INFO: generated fake credentials for clients4.google.com
INFO: generated fake credentials for clients1.google.com
INFO: generated fake credentials for localhost
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: socket hang up
    at createHangUpError (_http_client.js:215:15)
    at Socket.socketOnEnd (_http_client.js:300:23)
    at Socket.emit (events.js:129:20)
    at _stream_readable.js:908:16
    at process._tickCallback (node.js:355:11)
28 Aug 14:55:39 - [nodemon] app crashed - waiting for file changes before starting...
```
Error occurs when Chrome fails to load a page (https://debug.mms.local) and:
```
No data received
ERR_EMPTY_RESPONSE
```
error page shows up.
I added certificates to chrome. 

My code:
```javascript
var hoxy = require('hoxy');
var _ = require('lodash');
var fs = require('fs');

var PROXY_PORT = 8001;
var proxy = new hoxy.createServer({
	certAuthority: {
		key: fs.readFileSync('./my-private-root-ca.key.pem'),
		cert: fs.readFileSync('./my-private-root-ca.crt.pem')
	}
}).listen(PROXY_PORT, function() {
	console.log('proProxy listens on: ' + PROXY_PORT);
});

proxy.intercept({
	phase: 'response',
	url: '/api/mms/login',
	method: 'POST',
	mimeType: 'application/json',
	as: 'json'
}, function(req, resp, cycle) {
	console.log(resp.json.response);
});

proxy.intercept({
	phase: 'response',
	url: '/api/v1/login',
	method: 'POST',
	mimeType: 'application/json',
	as: 'json'
}, function(req, resp, cycle) {
	console.log(resp.json.response);
});
```
Could you help me with this?

**Comment by editedredx** [comment]
Can you reach https://debug.mms.local without the proxy? I think you're not running a server on that address or it's not responding. If I call https://some.local.address that's not directing anywhere listening I get the same error. So I'd say, try it without proxy first.

Note: the part that It's actually crashing shouldn't happen though.
Technically it should be returning a 504 Gateway Timeout.

**Comment by majdamarcin** [comment]
Yeah, i can reach it without proxy. I got apache server running on 127.0.0.1, it's mapped in hosts to debug.mms.local.

**Comment by greim** [comment]
I see `debug.mms.localhost` mentioned in one place and `debug.mms.local` in another place. Could there be a wrong url or misconfig somewhere?

**Comment by editedredx** [comment]
Regular https calls are working as expected, I think this is a different issue. I've run into something similar and created a new issue #60 for it. So I'm closing this one.
